### PR TITLE
Add bp energy mix dataset

### PR DESCRIPTION
**Pull request by pabloarosado**
I added the `etl` steps to generate the [BP energy mix dataset](https://owid.cloud/admin/datasets/5570).

A few comments:
* This dataset was previously done with [a script](https://github.com/owid/energy-data/blob/master/scripts/bp_energy.py) in the `energy-data` repos). I have now moved it to `etl`, made the code cleaner, and added plenty of documentation, especially to avoid confusion between direct and input-equivalent primary energy.
* One important improvement is that aggregate regions are now consistent with owid definitions. And all BP regions are kept (with names `* (BP)`).
* I included a 2021 dataset just to be able to fill missing data in the current dataset (which unfortunately has many gaps) with data point from the previous release. But this 2021 dataset is not used in grapher, it will only live in the catalog.

To be reviewed:
* @HannahRitchie could you please check the metadata file (`etl/steps/data/garden/bp/2022-07-14/energy_mix.meta.yml`)? This is what ends up in grapher.
* @Marigold could you please (just quickly, no need to look in detail) if the grapher step and the way I am handling metadata make sense?

Thanks!

### CODE REVIEW

**Comment by HannahRitchie** [comment]
Thanks @pabloarosado. I would suggest two small changes to the methodology text:

>  This input-equivalent primary energy takes account of the inefficiencies in fossil fuel production and provides a better
      approximation of each source's _share_ of "final energy" consumption.

[DELETE THESE TWO SENTENCES] 

>  For fossil fuels and biofuels, there is only direct primary energy.
> However, when calculating the share of fossil fuels to the total primary energy, this will be done both with respect to the total direct primary energy, and the total input-equivalent primary energy.

**Comment by pabloarosado** [comment]
> > @Marigold could you please (just quickly, no need to look in detail) if the grapher step and the way I am handling metadata make sense?
> 
> Yes, can't think of better way than yours.
> 
> Code looks great. Just one question - why do we have both `2022-07-11` and `2022-07-14` versions? There's a lot of code duplication so I'm wondering if it's worth DRYing some of the functions into a common module (reusable functions such as `load_population` could be even in `etl/data_utils.py`)

Thanks @Marigold . The two versions are for two versions of the dataset. The `2022-07-11` is for the 2021 release and the `2022-07-14` is for the 2022 release. So they need to be in different folders, and therefore they don't share code.

File: etl/steps/data/garden/bp/2022-07-11/energy_mix.py [hunk 1]
```diff
@@ -0,0 +1,794 @@
...
+                }
+            )
+            income_groups = pd.concat(
+                [income_groups, historic_region_df], ignore_index=True
+            )
+
+    return cast(pd.DataFrame, income_groups)
+
+
+def add_population(
```
[the remark this answers is not in the document]
Comment (pabloarosado) [reply etl/steps/data/garden/bp/2022-07-11/energy_mix.py:197]: Indeed, that function exists in datautils. But currently it does not include historical regions. That's why I copied and adapted the function here. I had to do the same thing in the FAOSTAT garden code. At some point we should work on datautils to improve this function.

**Comment by Marigold** [comment]
(Thanks, just wanted to be sure we want to keep both versions)

> So they need to be in different folders

You could also have module `data/garden/bp/shared.py` and in there e.g.
```
DIRECT_AND_EQUIVALENT_ENERGY = [
    "Hydro",
    ...
]
```
and then import from both versions
```
from ..shared import DIRECT_AND_EQUIVALENT_ENERGY
```
(plus you might need `__init__.py` in `data/garden/bp/` folder)

**Comment by Marigold** [comment]
> ...because those shared are likely to change over time.

Yeah, good call. Better than sharing across versions would be moving them to data-utils just like you said.

**Pull request merged**